src/Microsoft.AspNetCore.Blazor.Browser/Interop/JavaScriptInvoke.cs

```diff
@@ -0,0 +1,253 @@
+﻿using System;
+using System.Linq;
+using System.Reflection;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.Blazor.Browser.Interop
+{
+    /// <summary>
+    /// TODO
+    /// </summary>
+    public class JavaScriptInvoke
+    {
+        /// <summary>
+        /// TODO
+        /// </summary>
+        /// <param name="methodOptions"></param>
+        /// <param name="methodArguments"></param>
+        /// <returns></returns>
+        public static string InvokeDotnetMethod(string methodOptions, string methodArguments)
+        {
+            Console.WriteLine(methodOptions);
+            Console.WriteLine(methodArguments);
+            var options = JsonUtil.Deserialize<MethodOptions>(methodOptions);
+            var argumentDeserializer = GetOrCreateArgumentDeserializer(options);
+            var invoker = GetOrCreateInvoker(options, argumentDeserializer);
+
+            var result = invoker(methodArguments);
+            if (options.Async != null && !(result is Task))
+            {
+                throw new InvalidOperationException($"'{options.Method.Name}' in '{options.Type.TypeName}' must return a Task.");
+            }
+            if (result is Task taskResult)
+            {
+                Console.WriteLine(result.GetType().FullName);
+                taskResult.ContinueWith(task =>
+                {
+                    try
+                    {
+                        if (task.Status == TaskStatus.RanToCompletion)
+                        {
+                            if (task.GetType() == typeof(Task))
+                            {
+                                RegisteredFunction.Invoke<bool>(
+                                    options.Async.FunctionName,
+                                    options.Async.ResolveId);
+                            }
+                            else
+                            {
+                                var resultProperty = task.GetType()
```

Comment:
If we wanted to move the reflection up front (so that subsequent calls can extract the task result without using reflection), could use something like this: https://gist.github.com/SteveSandersonMS/c3a205f19264486753183f5b75c5d6b5

Reply:
Yes, I need to cleanup/cache things here